hw/mcu/stm/stm32l4xx/src/hal_flash.c

```diff
@@ -0,0 +1,167 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include <string.h>
+#include "stm32l4xx_hal_def.h"
+#include "stm32l4xx_hal_flash.h"
+#include "stm32l4xx_hal_flash_ex.h"
+#include "hal/hal_flash_int.h"
+#include "hal/hal_watchdog.h"
+
+static int stm32l4_flash_read(const struct hal_flash *dev, uint32_t address,
+        void *dst, uint32_t num_bytes);
+static int stm32l4_flash_write(const struct hal_flash *dev, uint32_t address,
+        const void *src, uint32_t num_bytes);
+static int stm32l4_flash_erase_sector(const struct hal_flash *dev,
+        uint32_t sector_address);
+static int stm32l4_flash_sector_info(const struct hal_flash *dev, int idx,
+        uint32_t *address, uint32_t *sz);
+static int stm32l4_flash_init(const struct hal_flash *dev);
+
+static const struct hal_flash_funcs stm32l4_flash_funcs = {
+    .hff_read = stm32l4_flash_read,
+    .hff_write = stm32l4_flash_write,
+    .hff_erase_sector = stm32l4_flash_erase_sector,
+    .hff_sector_info = stm32l4_flash_sector_info,
+    .hff_init = stm32l4_flash_init,
+};
+
+#define _FLASH_SIZE            (1024 * 1024)
+#define _FLASH_SECTOR_SIZE     2048
+
+const struct hal_flash stm32l4_flash_dev = {
+    .hf_itf = &stm32l4_flash_funcs,
+    .hf_base_addr = 0x08000000,
+    .hf_size = _FLASH_SIZE,
+    .hf_sector_cnt = _FLASH_SIZE / _FLASH_SECTOR_SIZE,
+    .hf_align = 8,
+};
+
+static int
+stm32l4_flash_read(const struct hal_flash *dev, uint32_t address, void *dst,
+        uint32_t num_bytes)
+{
+    memcpy(dst, (void *)address, num_bytes);
```

Comment:
It may be correct but in other cases address refers to hf_base_addr and in case of remap it could be wrong.

Reply:
Interesting! We don't do any remapping right now but if the user did, this would really break. This also applies to other flash drivers in the stm32 line (at least!).